src/net.rs

```diff
@@ -89,22 +92,143 @@ pub(crate) async fn connect_tcp(
         }
     }
 
-    let tcp_stream = match tcp_stream {
-        Some(tcp_stream) => tcp_stream,
-        None => {
-            return Err(
-                last_error.unwrap_or_else(|| format_err!("no DNS resolution results for {host}"))
-            );
+    Err(last_error.unwrap_or_else(|| format_err!("no DNS resolution results for {host}")))
+}
+
+pub(crate) async fn connect_tls(
+    context: &Context,
+    host: &str,
+    port: u16,
+    timeout_val: Duration,
+    strict_tls: bool,
+    alpns: &[&str],
+) -> Result<TlsStream<Pin<Box<TimeoutStream<TcpStream>>>>> {
+    let mut last_error = None;
+
+    for resolved_addr in
+        lookup_host_with_cache(context, host, port, timeout_val, strict_tls).await?
+    {
+        match connect_tls_inner(resolved_addr, timeout_val, host, strict_tls, alpns).await {
+            Ok(tls_stream) => {
+                if strict_tls {
+                    dns::update_connect_timestamp(context, host, &resolved_addr.ip().to_string())
+                        .await?;
+                }
+                return Ok(tls_stream);
+            }
+            Err(err) => {
+                warn!(context, "Failed to connect to {resolved_addr}: {err:#}.");
+                last_error = Some(err);
```

Comment:
Maybe better make it `first_error`, i.e. use `get_or_insert()` here? I think the first error is more interesting, it is the error from the recently used address (stale ones may be not working at all), moreover, we may hit some limit on connection attempts in this loop and such an error won't tell much